core/src/main/scala/org/apache/spark/rpc/netty/Dispatcher.scala

```diff
@@ -44,13 +45,26 @@ private[netty] class Dispatcher(nettyEnv: NettyRpcEnv, numUsableCores: Int) exte
   private val shutdownLatch = new CountDownLatch(1)
   private lazy val sharedLoop = new SharedMessageLoop(nettyEnv.conf, this, numUsableCores)
 
-  private def getMessageLoop(name: String, endpoint: RpcEndpoint): MessageLoop = {
-    endpoint match {
-      case e: IsolatedRpcEndpoint =>
-        new DedicatedMessageLoop(name, e, this)
-      case _ =>
-        sharedLoop.register(name, endpoint)
-        sharedLoop
+  private def assignToMessageLoop(
+      name: String,
+      endpoint: RpcEndpoint,
+      endpointRef: RpcEndpointRef): MessageLoop = {
+    // This must be done before assigning RpcEndpoint to MessageLoop, as MessageLoop sets Inbox be
+    // active when registering, and endpointRef must be put into endpointRefs before onStart is
+    // called. Refer the doc of `RpcEndpoint.self`, as well as `NettyRpcEnv.endpointRef`.
+    endpointRefs.put(endpoint, endpointRef)
```

Comment:
If we can update `endpointRefs` here, could we also update `endpointRefs` above `getMessageLoop`/`assignToMessageLoop`?

Reply:
Sorry I don't get it. Could you elaborate? If you meant creating endpointRef here, that would be simple to do but either we need to have tuple of return type or registerRpcEndpoint should get endpointRef from endpointRefs, looks like no big advantage.

Reply:
Ok, never mind. I got why you do this(`endpointRefs.put(endpoint, endpointRef)`) in `assignToMessageLoop()`.

Reply:
Yes, that's for doing only when putIfAbsent runs code for "absent".

Reply:
you read my mind.